### PR TITLE
Use single files for AP patches and widescreen cheats

### PR DESCRIPTION
**Pull request by Epicpkmn11**
#### What's changed?

- Makes AP patch and widescreen files all be combined into single files to save SD card space and extraction / copy time.
  - On a 32 KB cluster card, saves ~~about 22 MB~~ almost 40 MB now with widescreen!
  - Uses binary search of games so probably either faster or not much slower
  - Files are packed from separate files when building so it's still easy to add files. I also left in the old checks so you can easily add new AP fix files for testing and such
  - I just made a simple file format for this that's basically just a big sorted list of the games and where their data is, then all the data so it's easy to binary search and such. I used `.pck` (short for Pack) as it's a decent generic extension that's at least better than `.bin`

TODO:
- [x] Other themes and such
- [x] Test it better, both normal and cheat data seemed to work but I didn't test much
- [x] Maybe do this to widescreen too, though I still think that might be better in a usrcheat.dat format file

#### Where have you tested it?

- DSi (K) from Unlaunch
- n2DS XL (U) with latest Luma and such

*** 
#### Pull Request status
- [x]  This PR has been tested using the provided devkitPro, devkitARM, and EasyGL2D.

### CODE REVIEW

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 1]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+				fseek(file, offset, SEEK_SET);
+				u8 *buffer = new u8[size];
+				fread(buffer, 1, size, file);
+
+				mkdir("fat:/_nds", 0777);
+				mkdir("fat:/_nds/nds-bootstrap", 0777);
+				snprintf(ipsPath, sizeof(ipsPath), "%s:/_nds/nds-bootstrap/apFix%s", sdFound() ? "sd" : "fat", cheatVer ? "Cheat.bin" : ".ips");
+				FILE *out = fopen(ipsPath, "wb");
+				if(out) {
+					fwrite(buffer, 1, size, out);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:310]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp [hunk 2]
```diff
@@ -85,11 +85,43 @@ u32 getSDKVersion(FILE *ndsFile)
...
 
-	if (access(ipsPath, F_OK) == 0) {
-		return 0;
+			// Try binary search for the game
+			int left = 0;
+			int right = fileCount;
+
+			while (left <= right) {
+				int mid = left + ((right - left) / 2);
+				fseek(file, 16 + mid * 16, SEEK_SET);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp:109]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp [hunk 3]
```diff
@@ -85,11 +85,43 @@ u32 getSDKVersion(FILE *ndsFile)
...
+				fread(buf, 1, 4, file);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp:110]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp [hunk 4]
```diff
@@ -85,11 +85,43 @@ u32 getSDKVersion(FILE *ndsFile)
...
 {
-	char ipsPath[256];
-	snprintf(ipsPath, sizeof(ipsPath), "%s:/_nds/TWiLightMenu/apfix/%s-%X.ips", sdFound() ? "sd" : "fat", gameTid[num], headerCRC[num]);
+	FILE *file = fopen(sdFound() ? "sd:/_nds/TWiLightMenu/extras/apfix.pck" : "fat:/_nds/TWiLightMenu/extras/apfix.pck", "rb");
+	if (file) {
+		char buf[5] = {0};
+		fread(buf, 1, 4, file);
+		if (strcmp(buf, ".PCK") == 0) { // Make sure correct file type
+			u32 fileCount;
+			fread(&fileCount, 1, sizeof(fileCount), file);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/ndsheaderbanner.cpp:101]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 5]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+					right = mid - 1;
+				}
+			}
+
+			if (offset > 0 && size > 0) {
+				fseek(file, offset, SEEK_SET);
+				u8 *buffer = new u8[size];
+				fread(buffer, 1, size, file);
+
+				mkdir("fat:/_nds", 0777);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:305]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 6]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+				mkdir("fat:/_nds/nds-bootstrap", 0777);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:306]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 7]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+				fseek(file, 16 + mid * 16, SEEK_SET);
+				fread(buf, 1, 4, file);
+				nocashMessage(buf);
+				int cmp = strcmp(buf, gameTid[CURPOS]);
+				if (cmp == 0) { // TID matches, check CRC
+					u16 crc;
+					fread(&crc, 1, sizeof(crc), file);
+
+					if (crc == headerCRC[CURPOS]) { // CRC matches
+						fread(&offset, 1, sizeof(offset), file);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:284]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 8]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+						fread(&size, 1, sizeof(size), file);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:285]: Do we want to check the return value to detect errors?

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 9]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
...
+			int left = 0;
+			int right = fileCount;
+
+			while (left <= right) {
+				int mid = left + ((right - left) / 2);
+				fseek(file, 16 + mid * 16, SEEK_SET);
+				fread(buf, 1, 4, file);
+				nocashMessage(buf);
+				int cmp = strcmp(buf, gameTid[CURPOS]);
+				if (cmp == 0) { // TID matches, check CRC
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:279]: 0 => NULL

File: romsel_dsimenutheme/arm9/source/main.cpp [hunk 10]
```diff
@@ -253,6 +253,70 @@ std::string setApFix(const char *filename) {
 			return cheatVer ? "fat:/_nds/nds-bootstrap/apFixCheat.bin" : "fat:/_nds/nds-bootstrap/apFix.ips";
 		}
 		return ipsPath;
+	} else {
+		FILE *file = fopen(sdFound() ? "sd:/_nds/TWiLightMenu/extras/apfix.pck" : "fat:/_nds/TWiLightMenu/extras/apfix.pck", "rb");
+		if (file) {
+			char buf[5] = {0};
+			fread(buf, 1, 4, file);
```
Comment (eku) [inline romsel_dsimenutheme/arm9/source/main.cpp:260]: Do we want to check the return value to detect errors?

**Comment by Peter0x44** [comment]
This is great! I've wanted to see this change for a long time.
I will give it some testing myself.

File: romsel_dsimenutheme/arm9/source/main.cpp (hunk 1 above)
Comment (Epicpkmn11) [reply romsel_dsimenutheme/arm9/source/main.cpp:309]: Might be good I suppose, I don't think a single fread or fwrite in TWiLight is checked though so would go beyond just this change to make all everything safe

**Comment by Epicpkmn11** [comment]
Did the widescreen files too the same way, now it saves almost 40 MB of SD card space at 32KB clusters!

And did a bit more testing, DSi, R4, and DS classic menu all appear to be working correctly to me so I think this is good to merge.

**Comment by Epicpkmn11** [comment]
Oh does the resources Makefile work on Windows? I just did `./pack.py` which works fine on macOS and Linux as I have the bash thingy in the file, but that might not work on Windows

**Pull request merged**